UefiPayloadPkg/Library/BuildFdtLib/X86_BuildFdtLib.c

```diff
@@ -0,0 +1,980 @@
+
+      Data32      = (UINT32)((Data64 & 0x0FFFFFFFF));
+      RegData[13] = CpuToFdt32 (Data32);
+      DEBUG ((DEBUG_INFO, "PciRootBridge->MemAbove4G.size RegData[13] %x, \n", Data32));
+
+      // non-reloc/32bit/io, child-addr, parent-addr, length
+      Data32 = (N_NON_RELOCATABLE + SS_IO_SPACE);
+
+      RegData[14] = CpuToFdt32 (Data32);
+      DEBUG ((DEBUG_INFO, "PciRootBridge->Io.base RegData[14] %x, \n", Data32));
+
+      Data32 = (UINT32)PciRootBridgeInfo->RootBridge[Index].Io.Base;
+      // child-addr
+      RegData[15] = CpuToFdt32 (0);
+      RegData[16] = CpuToFdt32 (Data32);
+      DEBUG ((DEBUG_INFO, "PciRootBridge->Io.base RegData[16] %x, \n", Data32));
+
+      // parent-addr
+      RegData[17] = CpuToFdt32 (0);
+      RegData[18] = CpuToFdt32 (Data32);
+      // size
+      Data64 = (PciRootBridgeInfo->RootBridge[Index].Io.Limit - PciRootBridgeInfo->RootBridge[Index].Io.Base + 1);
+      if (Data64 & 0xFFFFFFFF00000000) {
+        Data32 = (UINT32)RShiftU64 ((Data64 & 0xFFFFFFFF00000000), 32);
+      } else {
+        Data32 = 0;
+      }
+
+      RegData[19] = CpuToFdt32 (Data32);
+      DEBUG ((DEBUG_INFO, "PciRootBridge->Io.base size [19] %x, \n", Data32));
+
+      Data32      = (UINT32)((Data64 & 0x0FFFFFFFF));
+      RegData[20] = CpuToFdt32 (Data32);
+      DEBUG ((DEBUG_INFO, "PciRootBridge->Io.base size [20] %x, \n", Data32));
+
+      Status = FdtSetProp (Fdt, TempNode, "ranges", &RegData, sizeof (RegData));
+      ASSERT_EFI_ERROR (Status);
+
+      // non-reloc/non-prefetch/memory, child-addr, parent-addr, length
+      // indicate rb1 does not support above 4GB DMA
+      Data32 = (N_NON_RELOCATABLE + P_NON_PREFETCHABLE + SS_32BIT_MEMORY_SPACE);
+
+      DMARegData[0] = CpuToFdt32 (Data32);
+      DEBUG ((DEBUG_INFO, "PciRootBridge->DMA base  RegData[0] %x, \n", Data32));
+
+      // child-addr
+      DMARegData[2] = CpuToFdt32 (0);
+      DMARegData[3] = CpuToFdt32 (0);
+      // parent-addr
+      DMARegData[4] = CpuToFdt32 (0);
+      DMARegData[5] = CpuToFdt32 (0);
+      // size
+      DMARegData[6] = CpuToFdt32 (1);
+      DMARegData[7] = CpuToFdt32 (0);
+
+      Status = FdtSetProp (Fdt, TempNode, "dma-ranges", &DMARegData, sizeof (DMARegData));
+      ASSERT_EFI_ERROR (Status);
+
+      Data32 = CpuToFdt32 (2);
+      Status = FdtSetProp (Fdt, TempNode, "#size-cells", &Data32, sizeof (UINT32));
+
+      Data32 = CpuToFdt32 (3);
+      Status = FdtSetProp (Fdt, TempNode, "#address-cells", &Data32, sizeof (UINT32));
+
+      BusNumber = PciRootBridgeInfo->RootBridge[Index].Bus.Base & 0xFF;
+      RegTmp[0] = CpuToFdt16 (BusNumber);
+      BusLimit  = PciRootBridgeInfo->RootBridge[Index].Bus.Limit & 0xFF;
+      RegTmp[1] = CpuToFdt16 (BusLimit);
+      DEBUG ((DEBUG_INFO, "PciRootBridge->BusNumber %x, \n", BusNumber));
+      DEBUG ((DEBUG_INFO, "PciRootBridge->BusLimit  %x, \n", BusLimit));
+
+      Status = FdtSetProp (Fdt, TempNode, "bus-range", &RegTmp, sizeof (RegTmp));

```

Comment:
2 elements in bus-range should be u32 type data. please help to update parser code together. (we still can "& 0xFF", but the data length should be u32)

Reply:
resolved.